### DISCUSSION TITLE
AD version of ComputeThermalExpansionEigenstrain with temperature dependent coefficient

### DISCUSSION THREAD

**Discussion by BoZeng1997** [idaholab/moose#23864] [Q&A Modules: Solid mechanics] [answered]
Hi,
I am trying to compute thermal expansion eigen strain $\epsilon=\alpha(T)(T-T_0)I$. I see there are two ways:

1.  In https://github.com/idaholab/moose/discussions/21785, `ComputeVariableEigenstrain` (I believe that is the new name for the object `ComputeVariableDependentEigenstrain`mentioned previously). The issue with this one is that it is non-AD. Do we have a AD version of it?
2. what I want is very close to what MOOSE already have, `Compute Thermal Expansion Eigenstrain`, where it implements $\epsilon=\alpha(T-T_0)I$. The adaptation seems not complicated since it supports AD already: re-define  `ADComputeThermalExpansionEigenstrain\thermal_expansion_coeff` to be `ADMaterialProperty<Real>` instead of `Real`, then use `ADParsedMaterial` to define such a material. 

I want to know which way is better. If 2 is better and no one is considering adding such feature, I can open an issue and submit a PR for it.

**Comment by lindsayad** [comment, marked as answer]
@bwspenc @dschwen @jiangwen84 do you have a suggestion for @BoZeng1997 ?

**Reply by jiangwen84** [reply]
Have you looked at ```ComputeMeanThermalExpansionFunctionEigenstrain```?  You can use temperature dependent thermal expansion.

**Reply by BoZeng1997** [reply]
@jiangwen84 Hi! Thanks for the recommendation. I am running simulation with `ComputeMeanThermalExpansionFunctionEigenstrain`.  From the description on the webage, `T_ref` is a required parameter and usually is different from stress free temperature. I could not find `T_ref` for my material's thermal expansion coefficient data. But I dont see a problem in my simulation when setting the same value for both parameters. What is your advice on it?

**Reply by jiangwen84** [reply]
In the second equation on the MOOSE documentation page (https://mooseframework.inl.gov/source/materials/ComputeMeanThermalExpansionFunctionEigenstrain.html#!), if you set T_sf = T_ref, then you basically get
epsilon = alpha (T-T_sf), which might be something you want.

In general, the T_sf and T_ref are different. You can also read this paper for more information https://www.sciencedirect.com/science/article/pii/S0029549311010387?ref=cra_js_challenge&fr=RR-1

**Reply by BoZeng1997** [reply]
thanks. that is very helpful!
